<a href="https://colab.research.google.com/github/MavelSterling/Informe_Inteligencia_Artificial/blob/main/notebook_Arboles.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ARBOLES DE DECISION EN DATOS DE SOBREVIVIENTES DEL TITANIC

In [7]:
#----------------------------------------------------------
# Se importan las librerias
#----------------------------------------------------------

import numpy as np
import pandas as pd
import sklearn
import warnings

## 1. Leer el archivo titanic.csv. Tenga en cuenta que los elementos del archivo están separados por ‘;’

In [8]:
#----------------------------------------------------------
# Url de los datos
#----------------------------------------------------------

url= ("https://raw.githubusercontent.com/MavelSterling/Informe_Inteligencia_Artificial/main/titanic.csv")

#----------------------------------------------------------
# Se cargan los datos
#----------------------------------------------------------
datos = pd.read_csv(url, sep=";")
datos

survived                                               Name  pclass  \
0           0                            Braund, Mr. Owen Harris       3   
1           1  Cumings, Mrs. John Bradley (Florence Briggs Th...       1   
2           1                             Heikkinen, Miss. Laina       3   
3           1       Futrelle, Mrs. Jacques Heath (Lily May Peel)       1   
4           0                           Allen, Mr. William Henry       3   
..        ...                                                ...     ...   
886         0                              Montvila, Rev. Juozas       2   
887         1                       Graham, Miss. Margaret Edith       1   
888         0           Johnston, Miss. Catherine Helen "Carrie"       3   
889         1                              Behr, Mr. Karl Howell       1   
890         0                                Dooley, Mr. Patrick       3   

        sex   age  sibsp  parch     fare    who  embark_town alone  
0      male  22.0      1      0   7.2500    man  Southampton    no  
1    female  38.0      1      0  71.2833  woman    Cherbourg    no  
2    female  26.0      0      0   7.9250  woman  Southampton   yes  
3    female  35.0      1      0  53.1000  woman  Southampton    no  
4      male  35.0      0      0   8.0500    man  Southampton   yes  
..      ...   ...    ...    ...      ...    ...          ...   ...  
886    male  27.0      0      0  13.0000    man  Southampton   yes  
887  female  19.0      0      0  30.0000  woman  Southampton   yes  
888  female   NaN      1      2  23.4500  woman  Southampton    no  
889    male  26.0      0      0  30.0000    man    Cherbourg   yes  
890    male  32.0      0      0   7.7500    man   Queenstown   yes  

[891 rows x 11 columns]

In [9]:
#----------------------------------------------------------
# Información de la base de datos
#----------------------------------------------------------

datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   Name         891 non-null    object 
 2   pclass       891 non-null    int64  
 3   sex          891 non-null    object 
 4   age          714 non-null    float64
 5   sibsp        891 non-null    int64  
 6   parch        891 non-null    int64  
 7   fare         891 non-null    float64
 8   who          891 non-null    object 
 9   embark_town  889 non-null    object 
 10  alone        891 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 76.7+ KB


## 2. Eliminar el atributo Name usando el comando datos.drop(["Name"], axis = 1, inplace = True)

In [10]:
#----------------------------------------------------------
# Eliminar el atributo Name
#----------------------------------------------------------

datos.drop(["Name"], axis = 1, inplace = True)

#----------------------------------------------------------
# Observar los datos sin el atributo Name
#----------------------------------------------------------
datos

survived  pclass     sex   age  sibsp  parch     fare    who  \
0           0       3    male  22.0      1      0   7.2500    man   
1           1       1  female  38.0      1      0  71.2833  woman   
2           1       3  female  26.0      0      0   7.9250  woman   
3           1       1  female  35.0      1      0  53.1000  woman   
4           0       3    male  35.0      0      0   8.0500    man   
..        ...     ...     ...   ...    ...    ...      ...    ...   
886         0       2    male  27.0      0      0  13.0000    man   
887         1       1  female  19.0      0      0  30.0000  woman   
888         0       3  female   NaN      1      2  23.4500  woman   
889         1       1    male  26.0      0      0  30.0000    man   
890         0       3    male  32.0      0      0   7.7500    man   

     embark_town alone  
0    Southampton    no  
1      Cherbourg    no  
2    Southampton   yes  
3    Southampton    no  
4    Southampton   yes  
..           ...   ...  
886  Southampton   yes  
887  Southampton   yes  
888  Southampton    no  
889    Cherbourg   yes  
890   Queenstown   yes  

[891 rows x 10 columns]

In [11]:
#----------------------------------------------------------
# Información de las variables de la base de datos
#----------------------------------------------------------
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   who          891 non-null    object 
 8   embark_town  889 non-null    object 
 9   alone        891 non-null    object 
dtypes: float64(2), int64(4), object(4)
memory usage: 69.7+ KB


In [12]:
#----------------------------------------------------------
# Variables categóricas (tipo object)
#----------------------------------------------------------
datos.select_dtypes(include=['object']).describe()

sex  who  embark_town alone
count    891  891          889   891
unique     2    3            3     2
top     male  man  Southampton   yes
freq     577  537          644   537

Se hace un resumen de las variables cuantitativas (tipo entero), para analizar la media, la desviación estándar, mínimo, 25%, 50% (mediana), 75% y el máximo

In [13]:
#----------------------------------------------------------
# Variables cuantitativas (tipo entero)
#----------------------------------------------------------
datos.select_dtypes(include=['int64']).describe()

survived      pclass       sibsp       parch
count  891.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642    0.523008    0.381594
std      0.486592    0.836071    1.102743    0.806057
min      0.000000    1.000000    0.000000    0.000000
25%      0.000000    2.000000    0.000000    0.000000
50%      0.000000    3.000000    0.000000    0.000000
75%      1.000000    3.000000    1.000000    0.000000
max      1.000000    3.000000    8.000000    6.000000

Se hace un resumen de las variables cuantitativas (tipo real), para analizar la media, la desviación estándar, mínimo, 25%, 50% (mediana), 75% y el máximo

In [14]:
#----------------------------------------------------------
# Variables cuantitativas (tipo real)
#----------------------------------------------------------
datos.select_dtypes(include=['float64']).describe()

age        fare
count  714.000000  891.000000
mean    29.699118   32.204208
std     14.526497   49.693429
min      0.420000    0.000000
25%     20.125000    7.910400
50%     28.000000   14.454200
75%     38.000000   31.000000
max     80.000000  512.329200

## 3. Seleccionar aleatoriamente el 80% del conjunto de datos para entrenar y el 20% restante para las pruebas

In [15]:
#----------------------------------------------------------
# Libreria sklearn
#----------------------------------------------------------
from sklearn.model_selection import train_test_split 

N=len(datos) # cantidad de datos
cTrain=int(N*0.8) # 80% para entrenar y 20% para probar
cTest=N-cTrain 

print(N,cTrain,cTest)# cantidad de: datos completos, datos entrenados, datos para probar

train_data,test_data= sklearn.model_selection.train_test_split(datos, train_size=cTrain, test_size=cTest)

891 712 179


891 es el número de observaciones en la base de datos

712 es el 80% de las observaciones de la base de datos que se van a utilizar para entrenar los datos

179 es el 20% de las observaciones de la base de datos que se van a utilizar para probar el modelo de predicción

In [16]:
#----------------------------------------------------------
# Dimensiones de los datos de entrenamiento
#----------------------------------------------------------
train_data.shape

(712, 10)

In [17]:
#----------------------------------------------------------
# Mostrar las primeras 5 filas de los datos de entrenamiento
#----------------------------------------------------------

train_data.head(5)

survived  pclass     sex   age  sibsp  parch     fare    who  \
465         0       3    male  38.0      0      0   7.0500    man   
184         1       3  female   4.0      0      2  22.0250  child   
394         1       3  female  24.0      0      2  16.7000  woman   
250         0       3    male   NaN      0      0   7.2500    man   
73          0       3    male  26.0      1      0  14.4542    man   

     embark_town alone  
465  Southampton   yes  
184  Southampton    no  
394  Southampton    no  
250  Southampton   yes  
73     Cherbourg    no

In [18]:
#----------------------------------------------------------
# Estadísticas descriptivas datos de entrenamiento
#----------------------------------------------------------
train_data.describe()

survived      pclass         age       sibsp       parch        fare
count  712.000000  712.000000  563.000000  712.000000  712.000000  712.000000
mean     0.380618    2.308989   29.789822    0.494382    0.389045   31.799425
std      0.485880    0.833563   14.260081    1.029780    0.815837   46.991133
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   21.000000    0.000000    0.000000    7.895800
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   30.548950
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

## 4. Utilizar una estrategia para normalizar los datos


### Pipeline para los atributos categóricos

In [19]:
#----------------------------------------------------------
#Libreria sklearn
#----------------------------------------------------------
from sklearn.preprocessing import OneHotEncoder 
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

# 5 variables categoricas
cat_attribs = ['pclass','sex','who','embark_town','alone']

cat_pipeline = Pipeline([
("imputer", SimpleImputer(strategy="most_frequent")), 
("cat_encoder", OneHotEncoder(sparse=False))
])


### Pipeline para los atributos numéricos

In [20]:
#----------------------------------------------------------
#Libreria sklearn
#----------------------------------------------------------
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler 
from sklearn.preprocessing import MinMaxScaler

# Variables númericas

num_attribs = ['pclass','age', 'fare', 'sibsp', 'parch']
num_pipeline = Pipeline([("imputer", SimpleImputer(strategy="median")),
                         ("scaler", StandardScaler())])

### Pipeline completo

In [21]:
#----------------------------------------------------------
#Libreria sklearn
#----------------------------------------------------------
from sklearn.compose import ColumnTransformer

full_pipeline = ColumnTransformer([ ("num", num_pipeline, num_attribs), 
                                   ("cat", cat_pipeline, cat_attribs),])

In [22]:
#----------------------------------------------------------
# Se transforman los datos entrenados 
# Variables independientes
#----------------------------------------------------------

X_train = full_pipeline.fit_transform(train_data)
X_train.shape

(712, 18)

In [23]:
X_train[0,:]

array([ 0.82956755,  0.67648889, -0.52705313, -0.48042264, -0.47720123,
        0.        ,  0.        ,  1.        ,  0.        ,  1.        ,
        0.        ,  1.        ,  0.        ,  0.        ,  0.        ,
        1.        ,  0.        ,  1.        ])

In [24]:
#----------------------------------------------------------
# Variable dependiente -> survived
#----------------------------------------------------------

y_train = train_data["survived"]
y_train

465    0
184    1
394    1
250    0
73     0
      ..
192    1
265    0
698    0
19     1
549    1
Name: survived, Length: 712, dtype: int64

### 5. Configurar los hiperparámetros del árbol de decisión de la siguiente manera: criterion=gini, splitter=best, y random_state=123. Obtener 10 árboles de decisión que resultan de modificar el hiperparámetro max_depth desde 10 hasta 100 con incrementos de 10

### 6. Incluya en el notebook una tabla con el accuracy para los 10 árboles del punto anterior

### 7. Repita el mismo procedimiento del punto 4 usando como hiperparámetros criterion=entropy, splitter=best, y random_state=123. Compare los 10 valores obtenidos con los del punto 5

### 8. Repita el mismo procedimiento del punto 4 usando como hiperparámetros criterion=entropy, splitter=random, y random_state=123.

### 9. Indique en el notebook los hiperparámetros que por el momento le permiten obtener el árbol con mayor accuracy

### 10. Seleccione uno de los hiperparámetros disponibles en la documentación (https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html) que sea diferente al criterion, splitter, max_depth, y random_state. Realice dos variaciones en el hiperparámetro seleccionado manteniendo los otros hiperparámetros del punto anterior. Indique el accuracy obtenido al modificar el hiperparámetro seleccionado y analice si el árbol de decisión mejora, empeora, o mantiene su exactitud.